domain/domain.module

```diff
@@ -95,3 +101,113 @@ function domain_preprocess_html(array &$variables) {
+ */
+function domain_confirm_fields($entity_type, $bundle, $text = array()) {
+  // We have reports that importing config causes this function to fail.
+
+  try {
+
+    $storage_id = $entity_type . '.' . DOMAIN_ADMIN_FIELD;
+    if (!$storage = \Drupal::entityTypeManager()->getStorage('field_storage_config')->load($storage_id)) {
+      $new_field_storage['id'] = $storage_id;
+      $new_field_storage['field_name'] = DOMAIN_ADMIN_FIELD;
+      $new_field_storage['type'] = 'entity_reference';
+      $new_field_storage['dependencies'] = ['domain' , 'user'];
+      $new_field_storage['entity_type'] = 'user';
+      $new_field_storage['cardinality'] = -1;
+      $new_field_storage['module'] = 'entity_reference';
+      $new_field_storage['settings']['target_type'] = 'domain';
+
+      $new_field_storage = FieldStorageConfig::create($new_field_storage);
+
+      $new_field_storage->save();
+    }
+
+    $id = $storage_id = $entity_type . '.' . $bundle . '.' . DOMAIN_ADMIN_FIELD;
+    if (!$field = \Drupal::entityTypeManager()->getStorage('field_config')->load($id)) {
+      $field = array(
+        'field_name' => DOMAIN_ADMIN_FIELD,
+        'entity_type' => $entity_type,
+        'label' => 'Domain administrator',
+        'bundle' => $bundle,
+        // Users should not be required to be a domain editor.
+        'required' => $entity_type !== 'user',
+        'description' => 'Select the domains this user may administer',
+        'default_value' => [],
+        'settings' => array(
+          'handler_settings' => array(
+            'sort' => array('field' => 'weight', 'direction' => 'ASC'),
+          ),
+        ),
+      );
+      $field_config = \Drupal::entityTypeManager()->getStorage('field_config')->create($field);
+      $field_config->save();
+    }
+    // Tell the form system how to behave. Default to radio buttons.
+    // @TODO: This function is deprecated, but using the OO syntax is causing
+    // test fails.
+    entity_get_form_display($entity_type, $bundle, 'default')
+      ->setComponent(DOMAIN_ADMIN_FIELD, array(
+        'type' => 'options_buttons',
+        'weight' => 50,
+      ))
+      ->save();
+  }
+  catch (Exception $e) {
+    \Drupal::logger('domain')->notice('Field installation failed.');
+  }
+}
+
+/**
+ * Implements hook_form_BASE_FORM_ID_alter() for \Drupal\user\UserForm.
+ *
+ * Handle settings that the user cannot access.
+ */
+function domain_form_user_form_alter(&$form, &$form_state, $form_id) {
+  // Add the options hidden from the user silently to the form.
+  $manager = \Drupal::service('domain.element_manager');
+  $form = $manager->setFormOptions($form, $form_state, DOMAIN_ADMIN_FIELD, TRUE);
+}
+
+/**
+ * Implements hook_domain_references_alter().
+ */
+function domain_domain_references_alter($query, $account, $context) {
+  // Restrict domains by assignment.
+  if ($context['field_type'] == 'admin' && $context['entity_type'] == 'user') {
+    if ($account->hasPermission('administer domains')) {
```

Comment:
Is this to prevent the case where a user has both 'administer domains' and 'assign domain administrators'?
If so, it would help to have a more explicit comment.